internal/mesh/internal/controllers/sidecarproxy/builder/destination_builder.go

```diff
@@ -0,0 +1,148 @@
+package builder
+
+import (
+	"github.com/hashicorp/consul/agent/connect"
+	"github.com/hashicorp/consul/internal/mesh/internal/types/intermediate"
+	pbcatalog "github.com/hashicorp/consul/proto-public/pbcatalog/v1alpha1"
+	pbmesh "github.com/hashicorp/consul/proto-public/pbmesh/v1alpha1"
+	"github.com/hashicorp/consul/proto-public/pbmesh/v1alpha1/pbproxystate"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+)
+
+func (b *Builder) BuildDestinations(destinations []*intermediate.Destination) *Builder {
+	for _, destination := range destinations {
+		if destination.Explicit != nil {
+			b.buildExplicitDestination(destination)
+		}
+	}
+
+	return b
+}
+
+func (b *Builder) buildExplicitDestination(destination *intermediate.Destination) *Builder {
+	clusterName := DestinationClusterName(destination.Explicit.DestinationRef, destination.Explicit.Datacenter, b.trustDomain)
+	statPrefix := DestinationStatPrefix(destination.Explicit.DestinationRef, destination.Explicit.Datacenter)
+
+	// All endpoints should have the same protocol as the endpoints controller ensures that is the case,
+	// so it's sufficient to read just the first endpoint.
+	if len(destination.ServiceEndpoints.Endpoints.Endpoints) > 0 {
+		// Get destination port so that we can configure this destination correctly based on its protocol.
+		destPort := destination.ServiceEndpoints.Endpoints.Endpoints[0].Ports[destination.Explicit.DestinationPort]
+
+		// Find the destination proxy's port.
+		// Endpoints refs will need to route to mesh port instead of the destination port as that
+		// is the port of the destination's proxy.
+		meshPortName := findMeshPort(destination.ServiceEndpoints.Endpoints.Endpoints[0].Ports)
+
+		if destPort != nil {
+			return b.addOutboundDestinationListener(destination.Explicit).
+				addRouter(clusterName, statPrefix, destPort.Protocol).
+				addCluster(clusterName, destination.Identities).
+				addEndpointsRef(clusterName, destination.ServiceEndpoints.Resource.Id, meshPortName)
+		}
+	}
+
+	return b
+}
+
+func (b *Builder) addOutboundDestinationListener(explicit *pbmesh.Upstream) *Builder {
+	listener := &pbproxystate.Listener{
+		Direction: pbproxystate.Direction_DIRECTION_OUTBOUND,
+	}
+
+	// Create outbound listener address.
+	switch explicit.ListenAddr.(type) {
```

Comment:
Is this a protobuf `oneof`? If so I thought there was a non-reflective way to manage this.